pkg/util/k8s/transform.go

```diff
@@ -0,0 +1,83 @@
+// Copyright 2024 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package k8s
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/client-go/tools/cache"
+)
+
+// NewTrimmer returns a cache.TransformFunc that can be used to trim objects stored in informers.
+// The function must be idempotent before client-go v0.31.0 to avoid a race condition happening when objects were
+// accessed during Resync operation, see https://github.com/kubernetes/kubernetes/issues/124337.
+// But it's generally more efficient to avoid trimming the same object more than once.
+func NewTrimmer(extraTrimmers ...cache.TransformFunc) cache.TransformFunc {
+	return func(obj interface{}) (interface{}, error) {
+		accessor, err := meta.Accessor(obj)
+		if err != nil {
+			return obj, nil
+		}
+		// It means the objects has been trimmed.
+		if accessor.GetManagedFields() == nil {
+			return obj, nil
+		}
+		// Trim common fields for all objects.
+		accessor.SetManagedFields(nil)
+		accessor.SetOwnerReferences(nil)
```

Comment:
Does that make using `Update` in our controllers dangerous? If these fields are present and are trimmed, and we later do an `Update` (not a `Patch`) in a controller, will the contents of these fields be lost?
I may be missing something but we do apply this trimmer to our CRDs. It is possible that some third-party software could integrate with Antrea, with some controllers that manage the lifecycle of Antrea CRDs. If these third-party controllers leverage owner references for lifecyle management (which is common and encouraged by kube-builder AFAIK), could he field be trimmed by the Antrea controller in storage, breaking garbage collection?

In other words, is it only safe to do this for resources that we do not `Update`?

Reply:
Good catch. I agree we shouldn't do this for all objects to avoid risks and complication. The considerable objects that may have the fields set we watch is Pod only, so it would be safer to move it to TrimPod.

Reply:
Sounds good.
I am also ok to do this for other objects if we ensure that we only ever use Patch, but then there is always a risk that we make a mistake in the future, so it should probably only be done if there is a noticeable reduction in memory usage.
For example, we could do it for Services if we replaced our usage of Update with Patch in the ServiceExternalIP controller.

Reply:
According to the test result, trimming owner reference is not that noticeable. I moved it to Pod specific trimmer and didn't create type specific trimmer for other types.